options/configurable_test.cc

```diff
@@ -0,0 +1,791 @@
+  ASSERT_EQ(*cp, 1);
+  ASSERT_EQ(*up, 1);
+  ASSERT_OK(c->ConfigureFromString(config_options_, "prepared=0"));
+  ASSERT_EQ(*up, 2);
+  ASSERT_EQ(*cp, 1);
+
+  ASSERT_NOK(c->ConfigureFromString(config_options_, "prepared=-2"));
+
+  c.reset(
+      new ValidatedConfigurable("Simple", TestConfigMode::kUniqueMode, true));
+  cp = c->GetOptions<int>("Prepared");
+  u = c->GetOptions<std::unique_ptr<Configurable>>("SimpleUnique");
+  up = u->get()->GetOptions<int>("Prepared");
+
+  ASSERT_OK(c->ConfigureFromString(config_options_, "prepared=0"));
+  ASSERT_EQ(*cp, 1);
+  ASSERT_EQ(*up, 0);
+}
+
+TEST_F(ConfigurableTest, DeprecatedOptionsTest) {
+  static std::unordered_map<std::string, OptionTypeInfo>
+      deprecated_option_info = {
+          {"deprecated",
+           {offsetof(struct TestOptions, b), OptionType::kBoolean,
+            OptionVerificationType::kDeprecated, OptionTypeFlags::kNone}}};
+  std::unique_ptr<Configurable> orig;
+  orig.reset(SimpleConfigurable::Create("simple", TestConfigMode::kDefaultMode,
+                                        &deprecated_option_info));
+  auto* opts = orig->GetOptions<TestOptions>("simple");
+  ASSERT_NE(opts, nullptr);
+  opts->d = true;
+  ASSERT_OK(orig->ConfigureOption(config_options_, "deprecated", "false"));
+  ASSERT_TRUE(opts->d);
+  ASSERT_OK(orig->ConfigureFromString(config_options_, "deprecated=false"));
+  ASSERT_TRUE(opts->d);
+}
+
+TEST_F(ConfigurableTest, AliasOptionsTest) {
+  static std::unordered_map<std::string, OptionTypeInfo> alias_option_info = {
+      {"bool",
+       {offsetof(struct TestOptions, b), OptionType::kBoolean,
+        OptionVerificationType::kNormal, OptionTypeFlags::kNone}},
+      {"alias",
+       {offsetof(struct TestOptions, b), OptionType::kBoolean,
+        OptionVerificationType::kAlias, OptionTypeFlags::kNone, 0}}};
+  std::unique_ptr<Configurable> orig;
+  orig.reset(SimpleConfigurable::Create("simple", TestConfigMode::kDefaultMode,
+                                        &alias_option_info));
+  auto* opts = orig->GetOptions<TestOptions>("simple");
+  ASSERT_NE(opts, nullptr);
+  ASSERT_OK(orig->ConfigureOption(config_options_, "bool", "false"));
+  ASSERT_FALSE(opts->b);
+  ASSERT_OK(orig->ConfigureOption(config_options_, "alias", "true"));
+  ASSERT_TRUE(opts->b);
+  std::string opts_str;
+  ASSERT_OK(orig->GetOptionString(config_options_, &opts_str));
+  ASSERT_EQ(opts_str.find("alias"), std::string::npos);
+
+  ASSERT_OK(orig->ConfigureOption(config_options_, "bool", "false"));
+  ASSERT_FALSE(opts->b);
+  ASSERT_OK(orig->GetOption(config_options_, "alias", &opts_str));
+  ASSERT_EQ(opts_str, "false");
+}
+
+TEST_F(ConfigurableTest, NestedUniqueConfigTest) {
+  std::unique_ptr<Configurable> simple;
+  simple.reset(
+      SimpleConfigurable::Create("Outer", TestConfigMode::kAllOptMode));
+  const auto outer = simple->GetOptions<TestOptions>("Outer");
+  const auto unique =
+      simple->GetOptions<std::unique_ptr<Configurable>>("OuterUnique");
```

Comment:
`OuterUnique` is not registered in the simple (outer is registered when calling Create), should unique be `nullptr`? I'm not clear here.

Reply:
In the SimpleConfigurable constructor around line 65 it registers the "OuterUnique" name.  It does this by taking the name of the object ("Outer") and concatenating it with "Unique".

Note that how the tests are registered and manipulated is way more complex than what the normal code is.  I tried to write the tests to use as few test classes to cover as many cases as possible and it created more complex tests than I would have liked.

Reply:
I see. Thanks for the explanation!